src/Val.h

```diff
@@ -222,90 +177,30 @@ class Val : public Obj {
 	IntrusivePtr<T> GetType() const
 		{ return cast_intrusive<T>(type); }
 
-#define CONST_ACCESSOR(tag, ctype, accessor, name) \
-	const ctype name() const \
-		{ \
-		CHECK_TAG(type->Tag(), tag, "Val::CONST_ACCESSOR", type_name) \
-		return val.accessor; \
-		}
-
-	// Needed for g++ 4.3's pickiness.
-#define CONST_ACCESSOR2(tag, ctype, accessor, name) \
-	ctype name() const \
-		{ \
-		CHECK_TAG(type->Tag(), tag, "Val::CONST_ACCESSOR", type_name) \
-		return val.accessor; \
-		}
-
-	CONST_ACCESSOR2(TYPE_BOOL, bool, int_val, AsBool)
-	CONST_ACCESSOR2(TYPE_INT, bro_int_t, int_val, AsInt)
-	CONST_ACCESSOR2(TYPE_COUNT, bro_uint_t, uint_val, AsCount)
-	CONST_ACCESSOR2(TYPE_DOUBLE, double, double_val, AsDouble)
-	CONST_ACCESSOR2(TYPE_TIME, double, double_val, AsTime)
-	CONST_ACCESSOR2(TYPE_INTERVAL, double, double_val, AsInterval)
-	CONST_ACCESSOR2(TYPE_ENUM, int, int_val, AsEnum)
-	CONST_ACCESSOR(TYPE_STRING, String*, string_val, AsString)
-	CONST_ACCESSOR(TYPE_FUNC, Func*, func_val, AsFunc)
-	CONST_ACCESSOR(TYPE_TABLE, PDict<TableEntryVal>*, table_val, AsTable)
-	CONST_ACCESSOR(TYPE_RECORD, std::vector<ValPtr>*, record_val, AsRecord)
-	CONST_ACCESSOR(TYPE_FILE, File*, file_val, AsFile)
-	CONST_ACCESSOR(TYPE_PATTERN, RE_Matcher*, re_val, AsPattern)
-	CONST_ACCESSOR(TYPE_VECTOR, std::vector<ValPtr>*, vector_val, AsVector)
-
-	const IPPrefix& AsSubNet() const
-		{
-		CHECK_TAG(type->Tag(), TYPE_SUBNET, "Val::SubNet", type_name)
-		return *val.subnet_val;
-		}
+#define UNDERLYING_ACCESSOR_DECL(ztype, ctype, name) \
+	ctype name() const;
+
+UNDERLYING_ACCESSOR_DECL(IntVal, bro_int_t, AsInt)
+UNDERLYING_ACCESSOR_DECL(BoolVal, bool, AsBool)
+UNDERLYING_ACCESSOR_DECL(EnumVal, int, AsEnum)
+UNDERLYING_ACCESSOR_DECL(CountVal, bro_uint_t, AsCount)
+UNDERLYING_ACCESSOR_DECL(DoubleVal, double, AsDouble)
+UNDERLYING_ACCESSOR_DECL(TimeVal, double, AsTime)
+UNDERLYING_ACCESSOR_DECL(IntervalVal, double, AsInterval)
+UNDERLYING_ACCESSOR_DECL(AddrVal, const IPAddr&, AsAddr)
+UNDERLYING_ACCESSOR_DECL(SubNetVal, const IPPrefix&, AsSubNet)
+UNDERLYING_ACCESSOR_DECL(StringVal, const String*, AsString)
+UNDERLYING_ACCESSOR_DECL(FuncVal, Func*, AsFunc)
+UNDERLYING_ACCESSOR_DECL(FileVal, File*, AsFile)
+UNDERLYING_ACCESSOR_DECL(PatternVal, const RE_Matcher*, AsPattern)
+UNDERLYING_ACCESSOR_DECL(TableVal, const PDict<TableEntryVal>*, AsTable)
```

Comment:
The goal is to not expose the underlying representations for records and vectors.  Ideally we wouldn't for tables, either, but that would require non-trivial work because right now the usual use for `AsTable()` is to iterate over the table's elements.  For records and vectors, I was able to add member functions to take care of the access idioms, but for tables these would be more involved (and not needed for the original goals motivating this change), in part due to the need to deal with `IterCookie`'s, which have to be exposed to the caller both for reentrancy and because it's appealing that `AsTable()` returns a `const` value.

That `VectorVal` has `UnderlyingValue` is vestigial from how I developed the change.  So my proposal is to remove that method and otherwise leave things as they are (exposing, at least for now, `AsTable()`).

Reply:
Ok, sounds reasonable.